docs/design/copy-tracking.md

```diff
@@ -10,17 +10,613 @@ results of rebasing across remote copies.
+name: baz, id: aaa111, copy_id: 4:baz->2:bar
+
+Merge commit:
+name: baz, id: aaa111, copy_id: 5:baz->{3:baz->1:foo,4:baz->2:bar}
+```
+
+We used the same content for both `foo` and `bar` above to simplify. If they
+had been different, we would have had a conflict in the contents but the copy
+ID would still have been clear.
+
+#### Example: Rebasing
+
+```
+$ jj log
+C rename bar->baz
+|
+B rename foo->bar
+|
+A add foo
+$ jj rebase -r C -d A
+```
+
+
+```
+$ jj log
+C rename foo->baz
+|
+| B rename foo->bar
+|/
+A add foo
+$ jj rebase -r C -d B
+```
+
+
+#### Example: Divergent renames
+
+Consider this "divergent rename" scenario:
+```
+$ jj log
+C rename foo->baz
+|
+| B rename foo->bar
+|/
+A add foo
+$ jj new B C
+```
+
+In this scenario, the regular 3-way merge of the trees without considering copy
+info results in a tree without conflicts. However, the user might reasonably
+expect to have to choose between the `bar` and `baz` names. Here's what Git says
+in this scenario:
+
+```
+$ git merge main
+CONFLICT (rename/rename): foo renamed to baz in HEAD and to bar in main.
+Automatic merge failed; fix conflicts and then commit the result.
+
+$ git st
+HEAD detached from ab0b8e3
+You have unmerged paths.
+  (fix conflicts and run "git commit")
+  (use "git merge --abort" to abort the merge)
+
+Unmerged paths:
+  (use "git add/rm <file>..." as appropriate to mark resolution)
+        added by them:   bar
+        added by us:     baz
+        both deleted:    foo
+```
+
+Interestingly, Git seems to represent this state by using index states that
+would not normally end up in the index as a result of conflicts.
+
+Here's what Mercurial says:
+
+```
+$ hg merge main
+note: possible conflict - foo was renamed multiple times to:
+ bar
+ baz
+1 files updated, 0 files merged, 0 files removed, 0 files unresolved
+(branch merge, don't forget to commit)
+```
+
+Mercurial doesn't have a place to record this state, so it just prints that
+note and leaves it at that.
+
+The model and algorithm described in this document would result in a conflict
+in the copy ID at both paths after propagating the renames.
+
+#### Example: @jonathantanmy's test case:
+
+TODO: fill this out
+
+```
+$ jj log
+E baz="baz" (resolves conflict)
+|
+D <conflict>
+|\
+C | rename bar->baz
+| |
+| B rename foo->baz
+|/
+A add foo="foo" and bar="bar"
+$ jj rebase -r E -d C
+$ jj rebase new D E -m F
+```
+
+If F is empty (auto-merged), it should have the same state as E before.
+
+
+### Log
+
+The copy graph contains all past paths and copy IDs of a file, so when doing
+`jj log <filename>`, we might want to translate that to a revset that's similar
+to `files()` but matches specific (path, copy ID) pairs instead of specific
+paths.
+
+### Annotate
+
+
+### Representation in Git
+
+Do we ever want to record renames in the Git backend? If we do, we would
+presumably store it outside the Git object, similar to how we store the change
+id for commits.
```

Comment:
Some thoughts, it'd be nice to support precise renames and conflicts in a best effort basis. But since the bar to entry here already is quite steep, which you point out nicely in the next paragraph, I'd rather go for a set of special commands to deal with it in the Git backend, when we communicate with a native backend.

Reply:
I think if we're going to index copy information for Git repos anyway, then it doesn't seem like much extra work to make it possible for a user to modify that state themselves.

Reply:
> I think if we're going to index copy information for Git repos anyway, then it doesn't seem like much extra work to make it possible for a user to modify that state themselves.

Yes, that's a good point. I think I previously over-indexed on the best effort use-case.